### ISSUE TITLE
Saving data into vuex coming from django-graphene backend.

### ISSUE THREAD

**Issue by andreisima** [vuex-orm/plugin-graphql#43]
Hi,

First of all thank you for vuex-orm. For me is just a dream come true to be able to replicate a part of back-end data to front-end and using vuex-orm-graphql to be able to do int in one request. I think this is a amazing project.

Now i try to use vuex-orm-graphql plugin, and i have some issues in implementing it in my project.
I am sending following query:
![screen shot 2018-07-30 at 14 36 15](https://user-images.githubusercontent.com/8919933/43395213-f3e4b502-9405-11e8-98aa-b93f79da2cba.png)

And i receive the following response:
![screen shot 2018-07-30 at 14 39 29](https://user-images.githubusercontent.com/8919933/43395342-6bd29d22-9406-11e8-9e6b-0354defb37ea.png)

Until this point all is as expected.

I would like to save a list of articles into store.
Now i need somebody to point me in the right direction because everything that i had tried failed.
My last "working" aka no errors saved in the store: 

![screen shot 2018-07-30 at 14 45 47](https://user-images.githubusercontent.com/8919933/43395611-5593e8e4-9407-11e8-9b0e-76743a381f67.png)

Also tried to create articleModel and articlesModel that has a many relation to articleModel. Failed.
I do receive this in console:

` Vuex-ORM: GraphQL Plugin  Ignoring entity article because it's not in the schema.`

I am doing something wrong and now i am out of ideas.

**Comment by phortx** [comment]
Thank you for the kind words, let's see what we can do for you.

Not sure what the actual issue is. Can you take a look in the Articles that come from the server?

A articles (plural) model shouldn't be necessary. All articles that come from the server are automatically saved in the store (or should at least).

Can you show us your Article model?

**Comment by andreisima** [comment]
Thank you for your fast response.

Below all files involved

// store/storeModels/ articleModel.js
```
import { Model } from '@vuex-orm/core'

export default class Article extends Model {
  static entity = 'article'

  static fields () {
    return {
      id: this.increment(),
      title: this.string('')
    }
  }
}
```
// store/storeModels/article.js
```
const defaultState = {}
const actions = {}
const mutations = {}
const getters = {}

export default {
  state: defaultState,
  getters,
  actions,
  mutations
}
```
// store/index.js
```
import Vue from 'vue'
import Vuex from 'vuex'

// Vue Orm
import VuexORM from '@vuex-orm/core'
import VuexORMGraphQL from '@vuex-orm/plugin-graphql'

// Models and data
import article from './storeData/article'
import articleModel from './storeModels/articleModel'

Vue.use(Vuex)

// VUE-ORM Setup
// Create new instance of Database.
const database = new VuexORM.Database()

// Register Model and Module. The First argument is the Model, and second is the Module.
database.register(articleModel, article)

// Auth headers for this apollo instance
const authHeaders = {
  'Authorization': 'JWT ' + localStorage.getItem('drft')
}

VuexORM.use(VuexORMGraphQL, {
  database: database,
  headers: authHeaders,
  url: 'http://93.114.129.7:8000/graphql',
  useGETForQueries: false,
  connectionQueryMode: 'edges',
  debug: process.env.NODE_ENV !== 'production'
})

export const store = new Vuex.Store({
  plugins: [
    VuexORM.install(database)
  ]
})
```
Thank you.

**Comment by phortx** [comment]
> `static entity = 'article'`

This should be in plural (`articles`), I think.

Everything else looks good for me. If this doesn't fix the issue, maybe the RAW response from the server would help, I think :)

**Comment by andreisima** [comment]
static entity = 'articles'
Same output

RAW output as a print screen from console
![screen shot 2018-07-30 at 15 45 35](https://user-images.githubusercontent.com/8919933/43398570-27f16ee8-9411-11e8-8c17-ac5762958183.png)

Also in this output there is a warning 

> vuex-orm-graphql.esm.js?71e0:21483  Vuex-ORM: GraphQL Plugin  Ignoring entity article because it's not in the schema.

Might be a issue... and i quite do not understand how is that not in schema.

**Comment by phortx** [comment]
That could be an issue. It means that there is no `article` query to fetch single a single article. This is required currently, I think.

**Comment by andreisima** [comment]
Ok, i will implement a model to retrieve only one item from backend and see from there.

**Comment by phortx** [comment]
You don't need a model, you just to ensure that there is a `article` query in your GraphQL Schema.

**Comment by andreisima** [comment]
In the backend i have an article query.
As you can see in the attached.
Also schema introspection looks ok.

![screen shot 2018-07-30 at 16 42 13](https://user-images.githubusercontent.com/8919933/43401100-fbca56ac-9417-11e8-9801-24957c7a6099.png)

**Comment by phortx** [comment]
Looks good. What I can see here is that the ID is not numeric. Someone told us in Slack yesterday that this doesn't work currently.

I can't tell for sure that this is the issue, but it could. I will try to discuss that with @kiaking

**Comment by andreisima** [comment]
I can confirm that the initial schema fetch is OK and there is a Query[0].article in there.
I can try to override this in the backend ... but i do not know if this is even possible and i if so i can risk brake some graphQL internals since that id is unique all records.

Off this topic.
Docs on vuex-orm-grapql are a bit misleading.
https://vuex-orm.github.io/vuex-orm-graphql/guide/fetch/#fetching-a-single-record
The query output is not like the one listed. Instead is (notice the extra argument filter):
```
query Articles($suid: Float!) {
  articles(filter: {suid: $suid}) {
    edges {
      node {
        id
      }
    }
  }
}
```

**Comment by andreisima** [comment]
> Looks good. What I can see here is that the ID is not numeric. Someone told us in Slack yesterday that this doesn't work currently.

I am not sure that this is the issue here.
I had read the code for vuex-orm and ID can be a string (in the master branch repo).
Now investigating this further i had found the following problem in my case.

In following function/class (this is still confusing me in JS):
```
Schema.prototype.getType = function (name, allowNull) {
    if (allowNull === void 0) { allowNull = false; }
    name = upcaseFirstLetter(name);
    var type = this.types.get(name);
    if (!allowNull && !type)
        throw new Error("Couldn't find Type of name " + name + " in the GraphQL Schema.");
    return type || null;
};
```
Here it is looking for the Type of the <object> directly in the schema. That is quite correct. 
Now my problem is naming.
Module is looking for <Object> and i have  <ObjectNode> in the schema. I will try to rename it in the backend and see if this is changing something.

OK.... this is working... Now i have to follow other 2 warnings.

vuex-orm-graphql.esm.js?71e0:21483  Vuex-ORM: GraphQL Plugin  Could find the argument with name id for the mutation/query articles
vuex-orm-graphql.esm.js?71e0:21483  Vuex-ORM: GraphQL Plugin  Could find the argument with name id for the mutation/query articles

Now i am 80% sure that all the problems are from naming convention that vuex-orm/graphql expects to be delivered from the server. This can be a problem in the long run. I do come from django/python world where the "Explicit over Implicit" is enforced, and this can cause some issues with graphql in the front-end. Maybe a better documentation on what is expected to delivered to module. I might help on that.

Will update this with more. Maybe somebody from django world will need it. Also i will edit the title to reflect the problem.

**Comment by andreisima** [comment]
@phortx 
Hi, 
**IMPORTANT : When i query for OBJECT it is saved in the store**
When i query for OBJECTS this is a different story.
I have managed to narrow everything down up to this:
```
Item
    $id: "_no_key_1"
    $isPersisted: true
    id: null
    title: ""
```
I bet that in Transformer.transformIncomingData the data arrives in a different format. And that i think is because my data gets out from here like this:
```
OJECTS: {
    edges [
        0: { node { id, other, other2}},
        1: {node {id, other, other2}},
        .....
    ]
}
```
and i bet that the function responsible with saving it wants something like:
```
OBJECTS: {
        0: { node { id, other, other2}},
        1: {node {id, other, other2}},
        .....
}
```

Is there any way i can control this?
Also here 
```
/**
     * Transforms a set of incoming data to the format vuex-orm requires.
```
Is there any way to find out how do you like the incoming data to be formatted like?

Than you.

**Comment by phortx** [comment]
Sorry I'm really busy these days. Thanks for your investigation. This is something I thought of too.

There are different ways in GraphQL how you structure your connections. You're using the `edges` mode which is not well tested currently. You just discovered that the transformer doesn't yet respect the connection mode.

I will fix this. If you want to help, you could provide a PR with a spec example for the transformer unit test which fails for your case. :)

**Comment by andreisima** [comment]
Thank you.

I also will continue this investigation further, right now i am hands deep in Transformer and in Store.insertData
Up so far i have managed to get 10 records inserted in the store but they are not as expected
```
> data:Object
   >  _no_key_1:Object
            $id:"_no_key_1"
            $isPersisted:false
    >  _no_key_2:Object
            $id:"_no_key_2"
            $isPersisted:false
.......

```
I will dig more into this.
As a info for you am am using [Facebook Relay specs](https://facebook.github.io/relay/graphql/connections.htm) because that is what my back-end follows. 
If i try to customise the output (that can be done) i am losing all the good stuff that is coming by default (first, last, cursor, etc. ), and i fell no joy in implementing them manually, if can be done.

I will try to provide a PR and maybe a fix for it. I am not that proficient in JS but i can try.

**Comment by andreisima** [comment]
@phortx 
### Made it work for case connectionQuerryMode: edges
only problem see below on pct 2 Queries

```
Transformer.transformIncomingData = function (data, model, mutation, recursiveCall) {
        var _this = this;
        if (mutation === void 0) { mutation = false; }
        if (recursiveCall === void 0) { recursiveCall = false; }
        var result = {};
        var context = Context.getInstance();
        if (!recursiveCall) {
            context.logger.group('Transforming incoming data');
            context.logger.log('Raw data:', data);
        }
        if (isArray(data)) {
            result = (data).map(function (d) {
              return _this.transformIncomingData(d, model, mutation, true);
            });
        }
        else {
// HERE IS WHAT I DID
  // will be passed to final result 
          var myResult = {}
          Object.keys(data).forEach(function(key) {
            if (data[key] !== undefined && data[key] !== null) {
  // looking for edges in object
  // reference pct 1. 
              if (data[key].edges) {
                var edgesData = data[key].edges
  // removing __typename from node 
  // this was causing issues on Store.insert
                edgesData = Object.keys(edgesData).forEach(function(key) {
          // removing (thank you for lodash)
                  var nodeData = _.omit(edgesData[key], '__typename')
         // creating record for myResult
                  myResult[key] = nodeData.node
                })
              }
            }
          })
  // passing myResult to result
          result = myResult
        }
// case is false because i do not touch recursive
        if (!recursiveCall) {
            context.logger.log('Transformed data:', result);
            context.logger.groupEnd();
        }
// THIS ONE NEEDS MORE DOCUMENTATION FOR ME TO UNDERSTAND
// I do not know what this property does.
// currently i save data with $isPersisted = false;
// might bite me latter
        else {
            result['$isPersisted'] = true;
        }
        // Make sure this is really a plain JS object. We had some issues in testing here.
        return clone(result);
    };
    return Transformer;
}());
```

### Now i want to express some conclusions that i had drawn reading/working on this module

1. Config arguments
When **VuexORMGraphQL** gets instantiated there is an **connectionQuerryMode** attribute. In my case **edges**. This attribute can be checked in Transformer class in order to transform data accordingly, and not to check data blindly. May i suggest a **customConnection** attribute where the user will be in charge with **data transformation**? If none of defaults apply to user case then, user can transform his own data to fit Store.insert requirement.

2. Queries
The module complains if on query OBJECTS argument id is not present. From my point of view this is a good idea. Now module can build:  query OBJECTS (id: someID). Until now all good.

case:  fetch one OBJECT:
```
OBJECT.fetch(id: someId)
```
this defaults to following query:
```
query QUERY($id: ID!) {
  OBJECT(id: $id) {
    id
    title
  }
}
```
taking into account that OBJECTS is checked for argument id is there any reason why above case does not defaults to:
```
query QUERY($id: ID!) {
  OBJECTS(id: $id) {
    id
    title
  }
}
```
Example:
![screen shot 2018-08-02 at 14 36 16](https://user-images.githubusercontent.com/8919933/43581407-8004a404-9661-11e8-87a9-85cf2d950c2a.png)

And in this case my code will work for both queries because both will return same object structure, only difference is that OBJECT will have only one node and OBJECTS will have x nodes

Also the presence of id field in query/model, from my point of view, must be mandatory. Now is not and is defaulting to '_no_key_1', '_no_key_2' .... '_no_key_x'.

Again thank you for this module, and will try to help you the best i can. As mentioned i come from python world and my methods are more pythonic in thinking than JavaScript oriented which i still consider a necessary evil :).

On Monday i will do a PR on this. Now i have to finish some stuff and i am quite short of time.
Please let me know what do you think so far, when time will allow you to do so.

Have a nice day.

**Comment by phortx** [comment]
I'm not sure if I understood everything, but some points:

- The plugin should support everything of the GraphQL Standard, so I don't see a customConnection option here
- The query `OBJECTS(id: $id)` doesn't make any sense, because querying a connection should use a filter argument. Also a ID is unique so only one record will be returned and not multiple ones.

> Also the presence of id field in query/model, from my point of view, must be mandatory. 

This is true for querying a single record, but not for a connection.

**Comment by WitteStier** [comment]
Hi,

First of all Thank you for VuexORM and this awesome plugin.
My GraphQL server also uses `edges` and `node`'s, so i also experienced the lack of support for the `edges` connectionMode in the transformer.

Im willing to fix this issue, but before i start i want to know if someone already is working on this issue.

**Comment by phortx** [comment]
I'm really busy these days and don't have any time to work on this plugin currently. And I didn't hear something from @andreisima for long time, so I don't think anyone is working on this currently. Feel free to implement this :)

Thank you for your support.

**Comment by phortx** [comment]
This is fixed by @WitteStier with #58 . Thank you very much!

I will release RC 17 now so you guys can test if this works for you.
